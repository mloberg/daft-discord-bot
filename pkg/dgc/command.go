package dgc

import "github.com/bwmarrin/discordgo"

// Command represents a Discord slash command
type Command struct {
	Name        string
	Description string
	Options     []*discordgo.ApplicationCommandOption

	Run func(s *discordgo.Session, i *discordgo.InteractionCreate) error
}
