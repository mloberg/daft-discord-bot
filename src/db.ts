import { PrismaClient } from '@prisma/client';

const client = new PrismaClient({
    // TODO: setup logging
});

export default client;
