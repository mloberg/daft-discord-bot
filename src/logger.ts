import fs from 'fs';
import pino from 'pino';

import config from './config';

const logger = pino(
    {
        level: config.logLevel,
        prettyPrint: config.debug,
    },
    config.isTest ? fs.createWriteStream('/dev/null') : process.stderr,
);

export default logger;
