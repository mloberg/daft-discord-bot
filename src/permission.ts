import { Guild, GuildMember } from 'discord.js';

import config from './config';

export const ensureRole = async (guild: Guild): Promise<void> => {
    const role = guild.roles.cache.find((role) => role.name === config.djRole);
    if (role) {
        return;
    }

    await guild.roles.create({ data: { name: config.djRole } });
};
