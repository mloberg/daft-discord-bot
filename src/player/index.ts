import { StreamDispatcher, StreamOptions, VoiceConnection } from 'discord.js';

import { Player as PlayerInterface } from '../types';
import LocalFile from './local';

export class Player implements PlayerInterface {
    private players: PlayerInterface[];

    constructor(...players: PlayerInterface[]) {
        this.players = players;
    }

    play(location: string, connection: VoiceConnection, options: StreamOptions = {}): Promise<StreamDispatcher> {
        const player = this.players.find((player) => player.supports(location));
        if (!player) {
            throw new Error(`Could not find player for ${location}`);
        }

        return player.play(location, connection, options);
    }

    getTitle(location: string): Promise<string | null> {
        const player = this.players.find((player) => player.supports(location));
        if (!player) {
            throw new Error(`Could not find player for ${location}`);
        }

        return player.getTitle(location);
    }

    supports(location: string): boolean {
        for (const player of this.players) {
            if (player.supports(location)) {
                return true;
            }
        }

        return false;
    }
}

export default new Player(new LocalFile());
