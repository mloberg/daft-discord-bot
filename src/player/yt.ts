import { StreamDispatcher, StreamOptions, VoiceConnection } from 'discord.js';
import ytdl from 'ytdl-core-discord';

import { Player } from '../types';

export default class YouTube implements Player {
    async play(location: string, connection: VoiceConnection, options: StreamOptions = {}): Promise<StreamDispatcher> {
        options.type = 'opus';
        return connection.play(await ytdl(location, { highWaterMark: 1 << 25 }), options);
    }

    async getTitle(location: string): Promise<string | null> {
        const info = await ytdl.getBasicInfo(location);
        return info.title;
    }

    supports(location: string): boolean {
        return ytdl.validateURL(location);
    }
}
