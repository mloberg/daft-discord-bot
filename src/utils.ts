export function env(key: string, _default: string): string {
    return process.env[key] ?? _default;
}
