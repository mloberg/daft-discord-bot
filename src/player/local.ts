import { exec } from 'child_process';
import { StreamDispatcher, StreamOptions, VoiceConnection } from 'discord.js';
import { createReadStream, existsSync } from 'fs';
import { extname } from 'path';
import { promisify } from 'util';

import logger from '../logger';
import { Player } from '../types';

const shell = promisify(exec);

export default class LocalFile implements Player {
    async play(location: string, connection: VoiceConnection, options: StreamOptions = {}): Promise<StreamDispatcher> {
        const ext = extname(location);

        if ('.ogg' === ext) {
            return connection.play(createReadStream(location), { ...options, type: 'ogg/opus' });
        }

        if ('.webm' === ext) {
            return connection.play(createReadStream(location), { ...options, type: 'webm/opus' });
        }

        return connection.play(location, options);
    }

    async getTitle(location: string): Promise<string | null> {
        const { stdout } = await shell(`ffprobe -v quiet -print_format json -show_format -show_streams "${location}"`);
        logger.debug({ process: 'ffprobe', input: location, stdout });

        return JSON.parse(stdout).format?.tags?.title || null;
    }

    supports(location: string): boolean {
        return existsSync(location);
    }
}
