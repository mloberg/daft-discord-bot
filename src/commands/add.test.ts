import { Client, Guild, Message, TextChannel } from 'discord.js';
import { mocked } from 'ts-jest/utils';

import db from '../db';
import { FriendlyError } from '../error';
import logger from '../logger';
import { hasPermission } from '../permission';
import player from '../player';
import command from './add';

const mocks = {
    react: jest.fn(),
    db: mocked(db, true),
    logError: mocked(logger.error),
    player: mocked(player),
    permission: mocked(hasPermission),
};

jest.mock('discord.js', () => ({
    Client: jest.fn(),
    Guild: jest.fn(),
    TextChannel: jest.fn(),
    Message: jest.fn().mockImplementation(() => ({
        member: true,
        react: mocks.react,
    })),
}));

jest.mock('../db');
jest.mock('../logger');
jest.mock('../player');
jest.mock('../permission');

describe('_add configuration', () => {
    it('should have basic command infomation', () => {
        expect(command.name).toEqual('add');
        expect(command.description).toEqual('Add a song');
        expect(command.usage).toEqual('[FILE] [...TAGS]');
    });

    it('should have no aliases', () => {
        expect(command.alias).toBeUndefined();
    });
});

describe('_add', () => {
    let message: Message;

    beforeEach(() => {
        mocks.react.mockClear();
        mocks.react.mockReturnThis();
        mocks.db.song.create.mockClear();
        mocks.logError.mockClear();
        mocks.player.supports.mockClear();
        mocks.player.getTitle.mockClear();
        mocks.permission.mockClear();

        const client = new Client();
        const guild = new Guild(client, {});
        const channel = new TextChannel(guild, {});
        message = new Message(client, {}, channel);
    });

    it('adds a file to the manager with a title', async () => {
        mocks.permission.mockReturnValue(true);
        mocks.player.supports.mockReturnValue(true);
        mocks.player.getTitle.mockResolvedValue('Testing');

        await command.run(message, { _: ['test.mp3', 'foo', 'bar'], $0: 'add' });
        expect(mocks.react).toBeCalledWith('ðŸŽµ');

        expect(mocks.db.song.create).toHaveBeenCalledTimes(1);
        expect(mocks.db.song.create).toHaveBeenCalledWith({
            data: {
                title: 'Testing',
                location: 'test.mp3',
                tags: {
                    create: [{ tag: 'foo' }, { tag: 'bar' }],
                },
            },
        });
    });

    it('adds a file to the manager without a title', async () => {
        mocks.permission.mockReturnValue(true);
        mocks.player.supports.mockReturnValue(true);
        mocks.player.getTitle.mockResolvedValue(null);

        await command.run(message, { _: ['notitle.mp3', 'foo', 'bar'], $0: 'add' });
        expect(mocks.react).toBeCalledWith('ðŸŽµ');

        expect(mocks.db.song.create).toHaveBeenCalledTimes(1);
        expect(mocks.db.song.create).toHaveBeenCalledWith({
            data: {
                title: null,
                location: 'notitle.mp3',
                tags: {
                    create: [{ tag: 'foo' }, { tag: 'bar' }],
                },
            },
        });
    });

    it('will throw an error if no file given', async () => {
        mocks.permission.mockReturnValue(true);

        try {
            await command.run(message, { _: [], $0: 'add' });
            fail('expected error to be thrown');
        } catch (err) {
            expect(err).toBeInstanceOf(FriendlyError);
            expect(err.message).toEqual('Invalid command usage: [song] [...tag]');
        }
    });

    it('will throw an error if no tags given', async () => {
        mocks.permission.mockReturnValue(true);

        try {
            await command.run(message, { _: ['test.mp3'], $0: 'add' });
            fail('expected error to be thrown');
        } catch (err) {
            expect(err).toBeInstanceOf(FriendlyError);
            expect(err.message).toEqual('Invalid command usage: [song] [...tag]');
        }
    });

    it('will throw an error if file is unsupported', async () => {
        mocks.permission.mockReturnValue(true);
        mocks.player.supports.mockReturnValue(false);

        try {
            await command.run(message, { _: ['none.mp3', 'foo', 'bar'], $0: 'add' });
            fail('expected error to be thrown');
        } catch (err) {
            expect(err).toBeInstanceOf(FriendlyError);
            expect(err.message).toEqual('I was unable to add that. Unsupported type.');
        }
    });

    it('will throw an error if cannot be added to database', async () => {
        const dbErr = new Error('duplicate');
        mocks.db.song.create.mockRejectedValue(dbErr);
        mocks.permission.mockReturnValue(true);
        mocks.player.supports.mockReturnValue(true);
        mocks.player.getTitle.mockResolvedValue(null);

        try {
            await command.run(message, { _: ['test.mp3', 'foo', 'bar'], $0: 'add' });
            fail('expected error to be thrown');
        } catch (err) {
            expect(mocks.logError).toHaveBeenCalledTimes(1);
            expect(mocks.logError).toHaveBeenCalledWith(dbErr);
            expect(err).toBeInstanceOf(FriendlyError);
            expect(err.message).toEqual('I was unable to add that song. Does it exist already?');
        }
    });

    it('will throw an error if user does not have role', async () => {
        mocks.permission.mockReturnValue(false);

        try {
            await command.run(message, { _: ['test.mp3', 'foo', 'bar'], $0: 'add' });
            fail('expected error to be thrown');
        } catch (err) {
            expect(err).toBeInstanceOf(FriendlyError);
            expect(err.message).toEqual('You do not have permission to do that.');
        }
    });
});
