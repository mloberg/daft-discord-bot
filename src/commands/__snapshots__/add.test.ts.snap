// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_add is a command 1`] = `
Object {
  "description": "Add a song",
  "examples": Array [
    "counting-the-cost.webm dark menacing epic",
  ],
  "name": "add",
  "run": [Function],
  "usage": "<file> [...tags]",
}
`;
